clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/InactiveTopicPartitionChecker.java

```diff
@@ -0,0 +1,221 @@
+   *
+   * <p>This method gracefully shuts down the scheduled executor service,
+   * waiting up to 5 seconds for running tasks to complete before forcing shutdown.
+   *
+   * @throws Exception if there's an error stopping the service
+   */
+  @Override
+  public void stopInner() throws Exception {
+    if (executorService != null && !executorService.isShutdown()) {
+      executorService.shutdown();
+      try {
+        if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
+          executorService.shutdownNow();
+        }
+      } catch (InterruptedException e) {
+        executorService.shutdownNow();
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Checks for inactive topic partitions and handles them appropriately.
+   *
+   * <p>This method is called periodically by the scheduled executor service and performs
+   * the following operations for each consumer:
+   *
+   * <ol>
+   *   <li><b>Detection:</b> Identifies currently inactive topic partitions by comparing
+   *       the per-topic-partition last successful poll timestamp against the configured threshold.
+   *       Skips topic partitions that are already paused to avoid redundant checks.</li>
+   *   <li><b>Resume:</b> Resumes previously paused topic partitions that are no longer inactive</li>
+   *   <li><b>Pause:</b> Pauses newly detected inactive topic partitions and adds them to the tracking set</li>
+   * </ol>
+   *
+   * <p>A topic partition is considered inactive if:
+   * <ul>
+   *   <li>The time since its last successful poll exceeds {@code inactiveTopicPartitionThresholdInMs}</li>
+   *   <li>It has never been successfully polled (lastSuccessfulPollTimestamp == -1)</li>
+   * </ul>
+   *
+   * <p>The method includes comprehensive error handling to ensure that exceptions
+   * don't break the periodic execution schedule.
+   *
+   * <p><b>Note:</b> This implementation uses per-topic-partition statistics from {@link ConsumptionTask#getPartitionStats(PubSubTopicPartition)}
+   * to get accurate activity information for each topic partition individually, rather than using a global
+   * last poll timestamp for the entire consumer.
+   */
+  void checkInactiveTopicPartition() {
+    try {
+      long currentTime = System.currentTimeMillis();
+
+      // Process each consumer: check for inactive topic-partitions, then pause/resume as needed
+      for (Map.Entry<SharedKafkaConsumer, ConsumptionTask> entry: getConsumerToConsumptionTask().entrySet()) {
+        SharedKafkaConsumer consumer = entry.getKey();
+        ConsumptionTask task = entry.getValue();
+        Set<PubSubTopicPartition> previouslyPausedTopicPartitions =
+            getConsumerToPausedTopicPartitionsMap().computeIfAbsent(consumer, x -> new HashSet<>());
+        Set<PubSubTopicPartition> assignedTopicPartitions = consumer.getAssignment();
+        Set<PubSubTopicPartition> currentlyInactiveTopicPartitions = new HashSet<>();
+
+        LOGGER.info(
```

Comment:
This will likely flood logs?

Reply:
Actually it is per consumer task per check period. but it is not very useful hence I removed it.